File: LabsInJava/Lab5/VkontakteRecycleVIew/app/src/main/java/com/example/vkrv/PostsListAdapter.java
package com.example.vkrv;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;

import java.util.List;

public class PostsListAdapter extends RecyclerView.Adapter<PostsListAdapter.PostsViewHolder> {

    private List<Posts> postsList;

    private @Nullable ItemClickListener listener;
    private @Nullable FragmentButtonListener fragmentButtonListener;
    private @Nullable FragmentLikeListener fragmentLikeListener;


    public PostsListAdapter(List<Posts> postsList,
                            @Nullable ItemClickListener listener,
                            @Nullable FragmentButtonListener fragmentButtonListener,
                            @Nullable FragmentLikeListener fragmentLikeListener)

    {   Posts.postsList = postsList;
        this.postsList = postsList;
        this.listener = listener;
        this.fragmentButtonListener = fragmentButtonListener;
        this.fragmentLikeListener = fragmentLikeListener;
    }

    @NonNull
    @Override
    public PostsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType)
    {
        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.item_posts, null, false);
        RecyclerView.LayoutParams params = new RecyclerView.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
        );
        view.setLayoutParams(params);
        return new PostsViewHolder(view);

    }

    @Override
    public void onBindViewHolder(@NonNull final PostsViewHolder holder, final int position) {
        final Posts posts = Posts.postsList.get(getItemViewType(position));
        Glide.with(holder.image.getContext()).load(posts.getImage()).into(holder.image);
        Glide.with(holder.logo.getContext()).load(posts.getLogo()).into(holder.logo);
//       holder.image.setImageResource(posts.getImage());
//       holder.logo.setImageResource(posts.getLogo());

        holder.author.setText(posts.getAuthor());
        holder.time.setText(posts.getTime());
        holder.description.setText(String.valueOf(posts.getDescription()));
        holder.btnLike.setText(posts.getLike()+"");
        holder.views.setText(String.valueOf(posts.getViews()));
        holder.btnComment.setText(String.valueOf(posts.getComment()));
        holder.btnRepost.setText(String.valueOf(posts.getRepost()));
        if(posts.isLiked()==true){
            holder.btnLike.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_favorite_black_24dp, 0, 0, 0); }
        else{
            holder.btnLike.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_favorite_border_black_24dp, 0, 0, 0);
            }
        holder.btnLike.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
//                if(fragmentButtonListener!=null){
                    if(posts.isLiked()){
                        holder.btnLike.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_favorite_border_black_24dp, 0, 0, 0);
                        fragmentButtonListener.myClick(posts, 2);
                        fragmentLikeListener.removeItemLike(posts);
                        posts.setLiked(false);
                        int num = Integer.parseInt(holder.btnLike.getText().toString());
                        holder.btnLike.setText(String.valueOf(num-1));
                    }
                    else
                    {
                        holder.btnLike.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_favorite_black_24dp, 0, 0, 0);
                        fragmentButtonListener.myClick(posts, 1);
                        posts.setLiked(true);
                        int num = Integer.parseInt(holder.btnLike.getText().toString());
                        holder.btnLike.setText(String.valueOf(num+1));
                    }
                }
//            }
        });
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (listener != null) {
                    listener.itemClick(position, posts);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return Posts.postsList.size();
    }


    public class PostsViewHolder extends RecyclerView.ViewHolder {
        ImageView logo;
        TextView author;
        ImageView image;
        TextView time;;
        TextView description;
        Button btnLike;
        Button btnComment;
        Button btnRepost;
        TextView views;


        public PostsViewHolder(@NonNull View itemView) {
            super(itemView);
            logo = itemView.findViewById(R.id.logo);
            author = itemView.findViewById(R.id.author);
            image = itemView.findViewById(R.id.image);
            time = itemView.findViewById(R.id.tvPostTime);
            description = itemView.findViewById(R.id.tvDescription);
            btnLike = itemView.findViewById(R.id.btnLike);
            btnComment = itemView.findViewById(R.id.btnComment);
            btnRepost = itemView.findViewById(R.id.btnRepost);
            views = itemView.findViewById(R.id.views);
        }
    }
    public int getItemViewType(int position) {
        return position;
    }

    interface ItemClickListener {
        void itemClick(int position, Posts item);}

    public interface FragmentButtonListener{
        void myClick(Posts posts, int option);}

    public interface FragmentLikeListener{
     void removeItemLike(Posts posts);}

    public void removeLike(Posts posts) {
        int p = Posts.postsList.indexOf(posts);
        posts.setLiked(false);
//        posts.setLike(R.drawable.ic_favorite_border_black_24dp);
        Posts.postsList.set(p, posts);
        postsList.set(p, posts);
        this.notifyItemChanged(p); }

}

File: LabsInKotlin/Lab5/app/src/main/java/com/example/lab5/PostsListAdapter.kt
package com.example.lab5

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import java.lang.String

class PostsListAdapter(
    postsList: MutableList<Posts>,
    itemClickListener: ItemClickListener?,
    fragmentButtonListener: FragmentButtonListener?,
    fragmentLikeListener: FragmentLikeListener?
) :
    RecyclerView.Adapter<PostsListAdapter.PostsViewHolder>() {
    var postsList: MutableList<Posts>
    private val listener: ItemClickListener?
    private val fragmentButtonListener: FragmentButtonListener?
    private val fragmentLikeListener: FragmentLikeListener?

    init {
        Posts.postsArrayList = postsList
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PostsViewHolder {
        val view = LayoutInflater
            .from(parent.context)
            .inflate(R.layout.item_posts, null, false)
        val params = RecyclerView.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        view.layoutParams = params
        return PostsViewHolder(view)
    }

    override fun onBindViewHolder(holder: PostsViewHolder, position: Int) {
        val posts: Posts = Posts.postsList.get(getItemViewType(position))
        Glide.with(holder.image.context).load(posts.getImage()).into(holder.image)
        Glide.with(holder.logo.context).load(posts.getLogo()).into(holder.logo)
        //       holder.image.setImageResource(posts.getImage());
//       holder.logo.setImageResource(posts.getLogo());
        holder.author.setText(posts.getAuthor())
        holder.time.setText(posts.getTime())
        holder.description.setText(String.valueOf(posts.getDescription()))
        holder.btnLike.setText(posts.getLike().toString() + "")
        holder.views.setText(String.valueOf(posts.getViews()))
        holder.btnComment.setText(String.valueOf(posts.getComment()))
        holder.btnRepost.setText(String.valueOf(posts.getRepost()))
        if (posts.isLiked() === true) {
            holder.btnLike.setCompoundDrawablesWithIntrinsicBounds(
                R.drawable.ic_favorite_black_24dp,
                0,
                0,
                0
            )
        } else {
            holder.btnLike.setCompoundDrawablesWithIntrinsicBounds(
                R.drawable.ic_favorite_border_black_24dp,
                0,
                0,
                0
            )
        }
        holder.btnLike.setOnClickListener { //                if(fragmentButtonListener!=null){
            if (posts.isLiked()) {
                holder.btnLike.setCompoundDrawablesWithIntrinsicBounds(
                    R.drawable.ic_favorite_border_black_24dp,
                    0,
                    0,
                    0
                )
                fragmentButtonListener!!.myClick(posts, 2)
                fragmentLikeListener!!.removeItemLike(posts)
                posts.setLiked(false)
                val num = holder.btnLike.text.toString().toInt()
                holder.btnLike.text = (num - 1).toString()
            } else {
                holder.btnLike.setCompoundDrawablesWithIntrinsicBounds(
                    R.drawable.ic_favorite_black_24dp,
                    0,
                    0,
                    0
                )
                fragmentButtonListener!!.myClick(posts, 1)
                posts.setLiked(true)
                val num = holder.btnLike.text.toString().toInt()
                holder.btnLike.text = (num + 1).toString()
            }
        }
        holder.itemView.setOnClickListener { listener?.itemClick(position, posts) }
    }

    override fun getItemCount(): Int {
        return Posts.postsList.size()
    }

    inner class PostsViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var logo: ImageView
        var author: TextView
        var image: ImageView
        var time: TextView
        var description: TextView
        var btnLike: Button
        var btnComment: Button
        var btnRepost: Button
        var views: TextView

        init {
            logo = itemView.findViewById(R.id.logo)
            author = itemView.findViewById(R.id.author)
            image = itemView.findViewById(R.id.image)
            time = itemView.findViewById(R.id.tvPostTime)
            description = itemView.findViewById(R.id.tvDescription)
            btnLike = itemView.findViewById(R.id.btnLike)
            btnComment = itemView.findViewById(R.id.btnComment)
            btnRepost = itemView.findViewById(R.id.btnRepost)
            views = itemView.findViewById(R.id.views)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }

    internal interface ItemClickListener {
        fun itemClick(position: Int, item: Posts?)
    }

    interface FragmentButtonListener {
        fun myClick(posts: Posts?, option: Int)
    }

    interface FragmentLikeListener {
        fun removeItemLike(posts: Posts?)
    }

    fun removeLike(posts: Posts) {
        val p: Int = Posts.postsList.indexOf(posts)
        posts.setLiked(false)
        //        posts.setLike(R.drawable.ic_favorite_border_black_24dp);
        Posts.postsList.set(p, posts)
        postsList[p] = posts
        this.notifyItemChanged(p)
    }

    init {
        Posts.postsList = postsList
        this.postsList = postsList
        this.listener = listener
        this.fragmentButtonListener = fragmentButtonListener
        this.fragmentLikeListener = fragmentLikeListener
    }
}
